### ISSUE TITLE
Reducing CPP

### ISSUE THREAD

**Issue by Ericson2314** [nix-windows/nix#5]
We have a lot of CPP at the moment, and I think Eelco wanted it reduced. Perhaps we leverage things like the new C++ STL filesystem API to reduce it? If possible, we could land such changes first, too.

**Comment by volth** [comment]
Do you plan to eventually merge it into Nix/Linux?

**Comment by Ericson2314** [comment]
Absolutely! I think that's the best shot to keep things maintainable, and I want things like linux<->windows remote building. I also help this make Nix easier to port to still more systems since the Unix stuff will need to be abstracted a bit.

**Comment by volth** [comment]
The interesting case would be heterogeneous `nixops` (`morph`, ... whatever) installation (Linux both Windows).

That means, a single `nix-instantiate` should be able to work with both Linux and Windows paths at once.
And that is not well compatible with `boost::filesystem` or another form like
```cpp
#ifdef WINDOWS
class Path {...}
#else
class Path {...}
#endif
```

it must be
```cpp
class LinuxPath {...}
class WindowsPath {...}
```
on both platforms

**Comment by Ericson2314** [comment]
Oh if Boost doesn't give us that....bummer. OTOH now that there is a `Path` vs `StorePath` distinction, that might make some interop earlier, e.g. if eval-time is restricted to paths which are in the form `StorePath` / option valid unicode relative path?
